File: egtest-processor/src/main/java/com/vocalabs/egtest/processor/junit/AllClassesWriter.java
package com.vocalabs.egtest.processor.junit;

import com.vocalabs.egtest.annotation.EgLanguage;
import com.vocalabs.egtest.processor.AnnotationCollector;
import com.vocalabs.egtest.processor.EgTestWriter;
import com.vocalabs.egtest.processor.MessageHandler;
import com.vocalabs.egtest.processor.Settings;
import com.vocalabs.egtest.processor.data.EgItem;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Map;

/** Build JUnit test source code; this is the code generator entry point. */
public class AllClassesWriter implements EgTestWriter {

    private final File directoryToFill;
    private final Settings.AlreadyExistsBehavior directoryExistsBehavior;
    private final EgLanguage defaultLanguage;

    public AllClassesWriter(EgLanguage defaultLanguage,
                            Settings.AlreadyExistsBehavior directoryExistsBehavior,
                            File directoryToFill) {
        this.directoryToFill = directoryToFill;
        this.directoryExistsBehavior = directoryExistsBehavior;
        this.defaultLanguage = defaultLanguage;
    }

    @Override
    public void write(AnnotationCollector annotationCollector) throws Exception {
        if (Settings.AlreadyExistsBehavior.FAIL.equals(directoryExistsBehavior) && directoryToFill.exists()) {
            annotationCollector.getMessageHandler()
                    .error("EgTest target directory exists ("+directoryToFill.getAbsolutePath()+")");
            return;
        }

        boolean didCreateDir = directoryToFill.mkdirs();
        if (Settings.AlreadyExistsBehavior.DELETE.equals(directoryExistsBehavior) &&  ! didCreateDir) {
            File[] files = directoryToFill.listFiles();
            if (files == null)
                throw new IOException("Location for writing EgTest source is not a directory: "+directoryToFill);
            for (File f: files) {
                deleteRecursive(f);
            }
        }

        MessageHandler messageHandler = annotationCollector.getMessageHandler();
        Map<String, List<EgItem<?>>> itemsByClassName = annotationCollector.getItemsByClassName();
        for (Map.Entry<String, List<EgItem<?>>> entry: itemsByClassName.entrySet()) {
            String className = entry.getKey();
            List<EgItem<?>> items = entry.getValue();
            Map<String,EgLanguage> languageForClassName = annotationCollector.getLanguageForClassName();
            ClassWriter.createFileSpec(languageForClassName, defaultLanguage, className, messageHandler, items)
                .writeTo(directoryToFill);
        }
    }

    private void deleteRecursive(File file) {
        File[] files = file.listFiles();
        if (files != null) {
            for (File f: files) {
                deleteRecursive(f);
            }
        }
        file.delete();
    }
}

File: egtest-processor/src/main/java/com/vocalabs/egtest/processor/junit/AllClassesWriter.kt
package com.vocalabs.egtest.processor.junit

import com.vocalabs.egtest.annotation.EgLanguage
import com.vocalabs.egtest.processor.AnnotationCollector
import com.vocalabs.egtest.processor.EgTestWriter
import com.vocalabs.egtest.processor.Settings

import java.io.File
import java.io.IOException

/** Build JUnit test source code; this is the code generator entry point.  */
class AllClassesWriter(private val defaultLanguage: EgLanguage,
                       private val directoryExistsBehavior: Settings.AlreadyExistsBehavior,
                       private val directoryToFill: File) : EgTestWriter {

    @Throws(Exception::class)
    override fun write(annotationCollector: AnnotationCollector) {
        if (Settings.AlreadyExistsBehavior.FAIL == directoryExistsBehavior && directoryToFill.exists()) {
            annotationCollector.messageHandler
                    .error("EgTest target directory exists (${directoryToFill.absolutePath})")
            return
        }

        val didCreateDir = directoryToFill.mkdirs()
        if (Settings.AlreadyExistsBehavior.DELETE == directoryExistsBehavior && !didCreateDir) {
            val files = directoryToFill.listFiles()
                    ?: throw IOException("Location for writing EgTest source is not a directory: $directoryToFill")
            files.forEach { f -> deleteRecursive(f) }
        }

        val messageHandler = annotationCollector.messageHandler
        val itemsByClassName = annotationCollector.itemsByClassName
        for ((className, items) in itemsByClassName) {
            val languageForClassName = annotationCollector.languageForClassName
            ClassWriter.createFileSpec(languageForClassName, defaultLanguage, className, messageHandler, items)
                    .writeTo(directoryToFill)
        }
    }

    private fun deleteRecursive(file: File) {
        file.listFiles()?.forEach { deleteRecursive(it) }
        file.delete()
    }
}
